### PR TITLE
Fast calculation of the Wilcoxon test (U) statistic.

### PR DESCRIPTION
**Pull request by LTLA**

### CODE REVIEW

File: src/fast_wilcox.cpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                }
+            }
+            std::sort(sortspace.begin(), sortspace.end());
+            last = target_id[i];
+        }
+
+        // Identifying the cells with TF expression + peak.
+        {
+            int tf_offset = exprs_p[tf_id[i]], tf_last = exprs_p[tf_id[i]+1];
+            int peak_offset = exprs_p[peak_id[i]], peak_last = exprs_p[peak_id[i]+1];
```
[the remark this answers is not in the document]
Comment (LTLA) [reply src/fast_wilcox.cpp:146]: agreed

File: src/fast_wilcox.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                less_than[c] += equal0[c];
+
+                double total_equal = equal0[c] + equal1[c];
+                ties[c] += total_equal * (total_equal * total_equal - 1); 
+            }
+
+            std::fill(equal0.begin(), equal0.end(), 0);
+            std::fill(equal1.begin(), equal1.end(), 0);
+
+        } else if (self_okay) {
```
[the remark this answers is not in the document]
Comment (LTLA) [reply src/fast_wilcox.cpp:80]: Yes, I think I just missed an `} else {` before line 82.

File: src/fast_wilcox.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        }
+
+        {
+            std::fill(okay_zeros.begin(), okay_zeros.end(), 0);
+            std::fill(okay_total.begin(), okay_total.end(), 0);
+            std::fill(notokay_zeros.begin(), notokay_zeros.end(), 0);
+            std::fill(notokay_total.begin(), notokay_total.end(), 0);
+
+            // Looping through the sortspace and creating cluster-specific sorted vectors.
```
Comment (LTLA) [inline src/fast_wilcox.cpp:176]: this comment is not relevant anymore.

File: src/fast_wilcox.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+            auto col_u = output_auc.column(i);
+            auto output_auc_ptr = static_cast<double*>(col_u.begin());
+            auto col_tie = output_ties.column(i);
+            auto output_tie_ptr = static_cast<double*>(col_tie.begin());
+            compute_auc(sortspace, notokay_zeros, okay_zeros, clusters.begin(), okay.data(), workspace, output_auc_ptr, output_tie_ptr);
+
+            full_okay_zeros[0] = std::accumulate(okay_zeros.begin(), okay_zeros.end(), 0);
+            full_notokay_zeros[0] = std::accumulate(notokay_zeros.begin(), notokay_zeros.end(), 0);
+            compute_auc(sortspace, full_okay_zeros, full_notokay_zeros, full_cluster.data(), okay.data(), full_workspace, output_auc_ptr + ngroups, output_tie_ptr + ngroups);
```
Comment (TomVuod) [inline src/fast_wilcox.cpp:205]: The positions of arguments `full_okay_zeros` and `full_notokay_zeros` should be interchanged.

Comment (LTLA) [reply src/fast_wilcox.cpp:205]: :+1:

**Pull request merged**